### PR TITLE
rewrite: Implement `uri query` operations

### PR DESCRIPTION
**Pull request by armadi1809**
@francislavoie This is my Initial take at #6096. This is still to be improved as only add, delete and set are implemented, but I wanted to get your initial feedback on the overall architecture of the solution before implementing the remaining operations.

### CODE REVIEW

**Comment by francislavoie** [comment]
Yeah, good start.Couple things I notice off the bat:

- I think it should support a block (i.e. braces) to do multiple operations with a single handler, similar to the header directive

- We need to apply the replacer on the query values (and maybe the keys too? Probably doesn't hurt). Users are usually not going to use static values, more often than not it's going to be dynamic using placeholders.

File: modules/caddyhttp/rewrite/rewrite.go [hunk 1]
```diff
@@ -470,5 +478,45 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+
+	// Names of query parameters to delete
+	Delete []string `json:"delete,omitempty"`
+}
+
+func (q *queryOps) do(r *http.Request, repl *caddy.Replacer) {
+	query := r.URL.Query()
+
+	for _, addParam := range q.Add {
+		key := repl.ReplaceAll(addParam.Key, "")
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:497]: I think maybe if the key is empty it should do nothing. I don't think it's valid to have an empty query key? Need to check the HTTP spec to make sure but I'm pretty sure it's invalid.

I think it might also be invalid to have `=` in the key, but maybe Go encodes that? Can cover with a test.

File: modules/caddyhttp/rewrite/rewrite.go [hunk 2]
```diff
@@ -470,5 +478,45 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+		val := repl.ReplaceAll(addParam.Val, "")
+		query[key] = append(query[key], val)
+	}
+
+	for _, deleteParam := range q.Delete {
+		param := repl.ReplaceAll(deleteParam, "")
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:503]: Same, maybe need to check for empty

File: modules/caddyhttp/rewrite/rewrite.go [hunk 3]
```diff
@@ -470,5 +478,45 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+		delete(query, param)
+	}
+
+	for _, setParam := range q.Set {
+		key := repl.ReplaceAll(setParam.Key, "")
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:508]: Same

File: modules/caddyhttp/rewrite/rewrite.go (hunk 1 above)
Comment (armadi1809) [reply modules/caddyhttp/rewrite/rewrite.go:497]: I just skipped the iteration if the key is empty. I also added a key with an `=` sign to the url in the test I wrote and it looks like go encodes that.

File: modules/caddyhttp/rewrite/rewrite.go (hunk 3 above)
Comment (armadi1809) [reply modules/caddyhttp/rewrite/rewrite.go:514]: Same as above.

File: modules/caddyhttp/rewrite/rewrite.go (hunk 2 above)
Comment (armadi1809) [reply modules/caddyhttp/rewrite/rewrite.go:506]: Same above.

**Comment by armadi1809** [comment]
@francislavoie, out of the list of operations in the issue, which ones you think we should add on top of what we have here at the moment?

**Comment by francislavoie** [comment]
I think rename would be good to have. The defaulting operation isn't necessary. Value replacement would also be good to have.

We don't have to implement those right away though, we can add them in a follow-up.

**Comment by armadi1809** [comment]
I jus added rename. We can add replace later as it looked a little bit more involved (especially if we want to use regex like for the headers). Please let me know if any modifications are still necessary in this PR.

**Review by mholt — CHANGES_REQUESTED** [review]
This is looking really good! :) Nice work. Just a couple nits from my first pass. Probably ready to move out of draft state.

**Comment by armadi1809** [comment]
@mholt Thanks for the round of review! All your feedback items should now be resolved. I also removed the draft status from the PR.

File: modules/caddyhttp/rewrite/rewrite.go [hunk 4]
```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
 	}
 }
 
+// queryOps describes the operations to perform on a query like add, delete and replace fields
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:482]:
```suggestion
// queryOps describes the operations to perform on query keys: add, set, rename and delete.
```

File: modules/caddyhttp/rewrite/rewrite.go [hunk 5]
```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+type queryOps struct {
+	// Adds query parameters; does not replace any existing header fields.
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:484]:
```suggestion
	// Adds query parameters; does not overwrite an existing query field,
	// and only appends an additional value for that key if any already exist.
```

File: modules/caddyhttp/rewrite/rewrite.go [hunk 6]
```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+	Add []queryOpsArguments `json:"add,omitempty"`
+
+	// Sets query parameters; replaces existing header fields.
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:487]:
```suggestion
	// Sets query parameters; overwrites a query key with the given value.
```

File: modules/caddyhttp/rewrite/rewrite.go [hunk 7]
```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+	Set []queryOpsArguments `json:"set,omitempty"`
+
+	// Renames query keys
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:490]:
```suggestion
	// Renames a query key from Key to Val, without affecting the value.
```

File: modules/caddyhttp/rewrite/rewrite.go [hunk 8]
```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+	Rename []queryOpsArguments `json:"rename,omitempty"`
+
+	// Names of query parameters to delete
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:493]:
```suggestion
	// Deletes a given query key by name.
```

File: modules/caddyhttp/rewrite/rewrite.go [hunk 9]
```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+		delete(query, key)
+	}
+
+	r.URL.RawQuery = query.Encode()
+}
+
+type queryOpsArguments struct {
+	// A key in the query string. Note that query string keys may appear multiple times.
+	Key string `json:"key,omitempty"`
+	Val string `json:"val,omitempty"`
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:542]:
```suggestion
	// The value for the given operation; for add and set, this is
	// simply the value of the query, and for rename this is the
	// query key to rename to.
	Val string `json:"val,omitempty"`
```

File: modules/caddyhttp/rewrite/rewrite.go [hunk 10]
```diff
@@ -89,6 +89,10 @@ type Rewrite struct {
 	// Performs regular expression replacements on the URI path.
 	PathRegexp []*regexReplacer `json:"path_regexp,omitempty"`
 
+	// Mutates the query string of the URI.
+
+	Query *queryOps `json:"query_operations,omitempty"`
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:94]: Shouldn't be a space in between, otherwise the comment won't be seen as a Godoc comment when generating the docs.

Also I think we should just shorten this to `query`, the `_operations` doesn't really add much to the config output. The fact that we're doing an "operation" is clear enough by the verbs like "add/set/rename/delete".

```suggestion
	// Mutates the query of the URI by applying a set of operations.
	Query *queryOps `json:"query,omitempty"`
```

File: modules/caddyhttp/rewrite/rewrite.go [hunk 11]
```diff
@@ -470,5 +479,68 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+	Rename []queryOpsArguments `json:"rename,omitempty"`
+
+	// Names of query parameters to delete
+	Delete []string `json:"delete,omitempty"`
+}
+
+func (q *queryOps) do(r *http.Request, repl *caddy.Replacer) {
+	query := r.URL.Query()
+
+	for _, addParam := range q.Add {
```
Comment (francislavoie) [inline modules/caddyhttp/rewrite/rewrite.go:500]: I think we should think about the order in which we apply the query operations, such that it's done in the most useful order.

I think that `delete` should probably always be applies last, because it's the "most destructive". That would allow a prior operation to copy a value from a particular key. For example you might have `?foo=bar` and you want to result in `?bar=foobar`, you might do "set bar foo{query.foo}" then "delete foo".

I think rename should probably be first, because then you'd expect to be able to use the new name immediately in subsequent operations.

And finally I think set should probably be before rename, because it overwrites a key anyway, so if you were to add then set in that order, only the set would remain, but doing set then add would result in both values existing.

Then I think we could probably use a few tests to make sure the above :point_up: work as expected.

WDYT @mholt ?

**Review by francislavoie — CHANGES_REQUESTED** [review]

File: modules/caddyhttp/rewrite/rewrite.go (hunk 11 above)
Comment (armadi1809) [reply modules/caddyhttp/rewrite/rewrite.go]: @francislavoie I re ordered the operations as suggested above and I also added tests to cover the different use cases you described.  I will leave this discussion open for now in case we decide to change anything.

**Comment by armadi1809** [comment]
All the discussions above (except for the order of operations) should now be resolved.

File: modules/caddyhttp/rewrite/rewrite.go [hunk 12]
```diff
@@ -470,5 +478,73 @@ func changePath(req *http.Request, newVal func(pathOrRawPath string) string) {
...
+	Add []queryOpsArguments `json:"add,omitempty"`
+
+	// Sets query parameters; overwrites a query key with the given value.
+	Set []queryOpsArguments `json:"set,omitempty"`
+
+	// Renames a query key from Key to Val, without affecting the value.
+	Rename []queryOpsArguments `json:"rename,omitempty"`
+
+	// Deletes a given query key by name.
+	Delete []string `json:"delete,omitempty"`
```
Comment (mholt) [inline modules/caddyhttp/rewrite/rewrite.go:494]: What if we put these fields in the order of the operations? So first Rename, then Set, then Add, then Delete?

I think I agree with the reasoning for that order.

Comment (armadi1809) [reply modules/caddyhttp/rewrite/rewrite.go]: The fields should be in the right order now.

**Review by mholt — APPROVED** [review]
I think this is good to go -- let's give it a shot! Nice work @armadi1809 :)

**Review by francislavoie — APPROVED** [review]
